### PR TITLE
Harmony 1.06 Port

### PR DESCRIPTION
**Pull request by nomis52**

### CODE REVIEW

**Comment by nomis52** [comment]
This is almost entirely autogenerated. The interesting bits that changed between 1.02 and 1.06 are:

- the reset function changed from SoftReset() to PLIB_RESET_SoftwareResetEnable()
- The structure of the USB device changed - the endpoint tables are now in system_init.c rather than bootloader_usb_descriptors.c

File: Bootloader/firmware/Bootloader.X/nbproject/configurations.xml [hunk 1]
```diff
@@ -247,13 +255,14 @@
...
     <Elem>../linker</Elem>
     <Elem>../../../common</Elem>
-    <Elem>/opt/microchip/harmony/v1_02</Elem>
     <Elem>../../../linker</Elem>
     <Elem>../src/system_config/number8</Elem>
     <Elem>../../../boardcfg/ethernet_sk2</Elem>
     <Elem>../../../boardcfg/number8</Elem>
     <Elem>../src/system_config/number1</Elem>
     <Elem>../../../boardcfg/number1</Elem>
+    <Elem>../../../../../opt/harmony/v1_06</Elem>
```
Comment (peternewman) [inline Bootloader/firmware/Bootloader.X/nbproject/configurations.xml]: Make absolute

File: Bootloader/firmware/Bootloader.X/nbproject/configurations.xml [hunk 2]
```diff
@@ -247,13 +255,14 @@
...
+    <Elem>../../../../../opt/microchip/harmony/v1_06</Elem>
```
Comment (peternewman) [inline Bootloader/firmware/Bootloader.X/nbproject/configurations.xml]: Again

File: Bootloader/firmware/Bootloader.X/nbproject/configurations.xml [hunk 3]
```diff
@@ -667,7 +821,7 @@
         <property key="enable-unroll-loops" value="false"/>
         <property key="exclude-floating-point" value="false"/>
         <property key="extra-include-directories"
-                  value="../src;../../../common;../../../boardcfg/ethernet_sk2;../src/system_config/ethernet_sk2;/opt/microchip/harmony/v1_02/framework;../src/system_config/ethernet_sk2/framework;../../../../../../../opt/microchip/harmony/v1_02/bsp/pic32mx_eth_sk2"/>
+                  value="../src;../../../common;../../../boardcfg/ethernet_sk2;../src/system_config/ethernet_sk2;/opt/microchip/harmony/v1_06/framework;../src/system_config/ethernet_sk2/framework;/opt/microchip/harmony/v1_06/framework;../../../../../../../opt/microchip/harmony/v1_06/bsp/pic32mx_eth_sk2"/>
```
Comment (peternewman) [inline Bootloader/firmware/Bootloader.X/nbproject/configurations.xml]: The last one is relative.

File: Bootloader/firmware/Bootloader.X/nbproject/configurations.xml [hunk 4]
```diff
@@ -1604,13 +1838,14 @@
...
-        <languageToolchainVersion>1.33</languageToolchainVersion>
+        <languageToolchainVersion>1.40</languageToolchainVersion>
         <platform>4</platform>
       </toolsSet>
       <compileType>
         <linkerTool>
           <linkerLibItems>
-            <linkerLibFileItem>/opt/microchip/harmony/v1_02/bin/framework/peripheral/PIC32MX675F512H_peripherals.a</linkerLibFileItem>
+            <linkerLibFileItem>/opt/microchip/harmony/v1_06/bin/framework/peripheral/PIC32MX675F512H_peripherals.a</linkerLibFileItem>
+            <linkerLibFileItem>/opt/microchip/harmony/v1_06/bin/framework/peripheral/PIC32MX675F512H_peripherals.a</linkerLibFileItem>
```
Comment (peternewman) [inline Bootloader/firmware/Bootloader.X/nbproject/configurations.xml]: Why is this present twice?

File: Bootloader/firmware/src/bootloader_usb_descriptors.c [hunk 5]
```diff
@@ -32,7 +33,7 @@
 // USB Device Layer Function Driver Registration Table
 // ----------------------------------------------------------------------------
 static const USB_DEVICE_FUNCTION_REGISTRATION_TABLE g_func_table[1] = {
-  /* Function 3 - The DFU Interface */
+  // Function 3 - The DFU Interface
```
Comment (peternewman) [inline Bootloader/firmware/src/bootloader_usb_descriptors.c]: Do we want to Doxygen this?

File: Bootloader/firmware/src/bootloader_usb_descriptors.c [hunk 6]
```diff
@@ -223,22 +224,16 @@ static const USB_DEVICE_MASTER_DESCRIPTOR usbMasterDescriptor = {
...
 const USB_DEVICE_INIT usbDevInitData = {
   .moduleInit = {SYS_MODULE_POWER_RUN_FULL},
-  .stopInIdle = false,
-  .suspendInSleep = false,
-  .endpointTable = g_endpoint_table,
   .registeredFuncCount = 1,
   .registeredFunctions = (USB_DEVICE_FUNCTION_REGISTRATION_TABLE*) g_func_table,
   .usbMasterDescriptor = (USB_DEVICE_MASTER_DESCRIPTOR*) &usbMasterDescriptor,
   .deviceSpeed = USB_SPEED_FULL,
+  .driverIndex = DRV_USBFS_INDEX_0,
```
Comment (peternewman) [inline Bootloader/firmware/src/bootloader_usb_descriptors.c]: We're now USB FS, I thought it was just a comm port?

File: Bootloader/firmware/src/system_config/number1/framework/system/ports/src/sys_ports_static.c [hunk 7]
```diff
@@ -64,8 +64,9 @@ void SYS_PORTS_Initialize(void)
     
     
     /* PORT D Initialization */
-    PLIB_PORTS_DirectionOutputSet( PORTS_ID_0, PORT_CHANNEL_D,  SYS_PORT_D_TRIS ^ 0xFFFF);
     PLIB_PORTS_OpenDrainEnable(PORTS_ID_0, PORT_CHANNEL_D, SYS_PORT_D_ODC);
+    PLIB_PORTS_Write( PORTS_ID_0, PORT_CHANNEL_D,  SYS_PORT_D_LAT);
+	PLIB_PORTS_DirectionOutputSet( PORTS_ID_0, PORT_CHANNEL_D,  SYS_PORT_D_TRIS ^ 0xFFFF);
```
Comment (peternewman) [inline Bootloader/firmware/src/system_config/number1/framework/system/ports/src/sys_ports_static.c]: Fix the spacing please.

File: Bootloader/firmware/src/system_config/number8/framework/system/ports/src/sys_ports_static.c [hunk 8]
```diff
@@ -64,8 +64,9 @@ void SYS_PORTS_Initialize(void)
     
     
     /* PORT D Initialization */
-    PLIB_PORTS_DirectionOutputSet( PORTS_ID_0, PORT_CHANNEL_D,  SYS_PORT_D_TRIS ^ 0xFFFF);
     PLIB_PORTS_OpenDrainEnable(PORTS_ID_0, PORT_CHANNEL_D, SYS_PORT_D_ODC);
+    PLIB_PORTS_Write( PORTS_ID_0, PORT_CHANNEL_D,  SYS_PORT_D_LAT);
+	PLIB_PORTS_DirectionOutputSet( PORTS_ID_0, PORT_CHANNEL_D,  SYS_PORT_D_TRIS ^ 0xFFFF);
```
Comment (peternewman) [inline Bootloader/firmware/src/system_config/number8/framework/system/ports/src/sys_ports_static.c]: Again

**Comment by peternewman** [comment]
Generally LGTM, just a few comments.

File: Bootloader/firmware/Bootloader.X/nbproject/configurations.xml (hunk 1 above)
Comment (nomis52) [inline Bootloader/firmware/Bootloader.X/nbproject/configurations.xml]: Done

File: Bootloader/firmware/Bootloader.X/nbproject/configurations.xml (hunk 2 above)
Comment (nomis52) [inline Bootloader/firmware/Bootloader.X/nbproject/configurations.xml]: Done

File: Bootloader/firmware/Bootloader.X/nbproject/configurations.xml (hunk 3 above)
Comment (nomis52) [inline Bootloader/firmware/Bootloader.X/nbproject/configurations.xml]: Fixed

File: Bootloader/firmware/Bootloader.X/nbproject/configurations.xml (hunk 4 above)
Comment (nomis52) [inline Bootloader/firmware/Bootloader.X/nbproject/configurations.xml]: Not sure, the tools aren't that smart. Removed.

File: Bootloader/firmware/src/bootloader_usb_descriptors.c (hunk 5 above)
Comment (nomis52) [inline Bootloader/firmware/src/bootloader_usb_descriptors.c]: No, this isn't an external api

File: Bootloader/firmware/src/bootloader_usb_descriptors.c (hunk 6 above)
Comment (nomis52) [inline Bootloader/firmware/src/bootloader_usb_descriptors.c]: The name is wrong (chalk that up to another harmony oddness). USBFS is actually the low level USB driver. not a filesystem as one would imagine.

File: Bootloader/firmware/src/system_config/number1/framework/system/ports/src/sys_ports_static.c (hunk 7 above)
Comment (nomis52) [inline Bootloader/firmware/src/system_config/number1/framework/system/ports/src/sys_ports_static.c]: This is auto-gened so if I change it we introduce diffs in the harmony plugin.

File: Bootloader/firmware/src/system_config/number8/framework/system/ports/src/sys_ports_static.c (hunk 8 above)
Comment (nomis52) [inline Bootloader/firmware/src/system_config/number8/framework/system/ports/src/sys_ports_static.c]: Same here

File: Bootloader/firmware/src/bootloader_usb_descriptors.c (hunk 5 above)
Comment (peternewman) [inline Bootloader/firmware/src/bootloader_usb_descriptors.c]: Okay

File: Bootloader/firmware/src/bootloader_usb_descriptors.c (hunk 6 above)
Comment (peternewman) [inline Bootloader/firmware/src/bootloader_usb_descriptors.c]: Ah, I did wonder when I'd first looked at 1.06 when I started on Ja Rule.

File: Bootloader/firmware/src/system_config/number1/framework/system/ports/src/sys_ports_static.c (hunk 7 above)
Comment (peternewman) [inline Bootloader/firmware/src/system_config/number1/framework/system/ports/src/sys_ports_static.c]: Okay

File: Bootloader/firmware/src/system_config/number8/framework/system/ports/src/sys_ports_static.c (hunk 8 above)
Comment (peternewman) [inline Bootloader/firmware/src/system_config/number8/framework/system/ports/src/sys_ports_static.c]: Okay

**Comment by peternewman** [comment]
LGTM

**Pull request merged**